HW1/P4/P4.py

```diff
@@ -0,0 +1,99 @@
+from pyspark import SparkContext
+from P4_bfs import BFS
+import time
+
+def split_up_input(line):
+    """ Takes in a string, assumed to be of the form '"character", "issue"'
+    and returns a tuple of (character, issue)"""
+
+    split_line = line.split('"')
+    character = split_line[1]
+    issue = split_line[3]
+
+    return (character, issue)
+
+def get_all_individual_pairs(issue_list_tup):
+    """ Takes a list of characters, all of which are assumed to appear in 
+    the same issue, and returns all combinations of (char1, char2). This is
+    sort of like a really poorly thought out adjacency matrix. For use in conjunction with flatMap
+    and to be post-processed into an adjacency list representation.
+    """
+
+    list_of_chars = issue_list_tup[1]
+
+    edges = []
+    for c1 in list_of_chars:
+        for c2 in list_of_chars:
+            if (c1 != c2):              # avoid self-links 
+                edges.append((c1, c2))
+                edges.append((c2, c1))
+
+    return tuple(edges)                 # flatMap expects a tuple
+
+if __name__ == "__main__":
+
+    start_time = time.time()
+
+    # First get a SparkContext object
+    sc = SparkContext('local', 'BFS')
+    sc.setLogLevel('WARN')
+
+    # Now load in the data
+    data = sc.textFile('source.csv', 16)
+
+    # Get rid of the stupid quotes
+    chars_and_issues = data.map(split_up_input)
+
+    # Flip it around for a groupByKey 
+    issues_and_chars = chars_and_issues.map(lambda (x, y): (y, x))
+
+    # Now we have the issues as keys - all things with the same issue should be connected
+    issues_and_chars = issues_and_chars.groupByKey()
+
+    # Now we access every single edge that appears in the graph
+    all_edges = issues_and_chars.flatMap(get_all_individual_pairs)
+
+    # Now each key is a character, and it appears many times -
+    # once for each edge. Consolidate all of these into an adjacency
+    # list representation of the graph.
+    # We are on 4 cores so we give it 16 partitions. We will try to preserve this partitioning
+    # as we progress. We did not worry about the partitioning scheme of the previous RDD's
+    # because they were so transient.
+    marvel_graph = all_edges.groupByKey().partitionBy(16)
```

Comment:
Good job setting a sensible partition size and preserving it throughout the implementation! Reduces shuffle costs and definitely speeds up the code!